### PR TITLE
Check if the archive has already been added to avoid duplicates

### PR DESCRIPTION
**Pull request by petrhosek**
This avoids adding archives multiple times, which results in duplicate
objects in the resulting rlib, leading to symbol collision and link
failures. This could happen when crate contains multiple link attributes
that all reference the same archive.

### CODE REVIEW

**Comment by petrhosek** [comment]
cc @cramertj @tmandry

File: src/librustc_codegen_llvm/back/archive.rs [hunk 1]
```diff
@@ -36,11 +36,21 @@ enum Addition {
...
+        path: PathBuf,
         archive: ArchiveRO,
         skip: Box<dyn FnMut(&str) -> bool>,
     },
 }
 
+impl Addition {
+    fn path(&self) -> &PathBuf {
+        match self {
+            Addition::File { path, .. } => path,
```
Comment (Centril) [inline src/librustc_codegen_llvm/back/archive.rs]:
```suggestion
            Addition::File { path, .. } | Addition::Archive { path, .. } => path,
```

File: src/librustc_codegen_llvm/back/archive.rs [hunk 2]
```diff
@@ -188,12 +198,16 @@ impl<'a> LlvmArchiveBuilder<'a> {
                       -> io::Result<()>
         where F: FnMut(&str) -> bool + 'static
     {
-        let archive = match ArchiveRO::open(archive) {
+        let archive_ro = match ArchiveRO::open(archive) {
             Ok(ar) => ar,
             Err(e) => return Err(io::Error::new(io::ErrorKind::Other, e)),
         };
+        if let Some(_) = self.additions.iter().find(|&ar| ar.path() == archive) {
```
Comment (cramertj) [inline src/librustc_codegen_llvm/back/archive.rs]:
```suggestion
        if self.additions.iter().any(|&ar| ar.path() == archive) {
```

**Comment by cramertj** [comment]
r=me with nits fixed.

File: src/librustc_codegen_llvm/back/archive.rs [hunk 3]
```diff
@@ -36,11 +36,20 @@ enum Addition {
...
     },
     Archive {
+        path: PathBuf,
         archive: ArchiveRO,
         skip: Box<dyn FnMut(&str) -> bool>,
     },
 }
 
+impl Addition {
+    fn path(&self) -> &PathBuf {
```
Comment (lzutao) [inline src/librustc_codegen_llvm/back/archive.rs]: `-> &Path`

File: src/librustc_codegen_llvm/back/archive.rs [hunk 4]
```diff
@@ -243,7 +256,7 @@ impl<'a> LlvmArchiveBuilder<'a> {
                         strings.push(path);
                         strings.push(name);
                     }
-                    Addition::Archive { archive, skip } => {
+                    Addition::Archive { path: _, archive, skip } => {
```
Comment (lzutao) [inline src/librustc_codegen_llvm/back/archive.rs]: `Addition::Archive { archive, skip, .. }`

**Pull request merged**